File: tools/haskell/src/STUNIR/IR.hs
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : STUNIR.IR
Description : IR processing utilities
Copyright   : (c) STUNIR Team, 2026
License     : MIT
-}

module STUNIR.IR
  ( parseIR
  , irToJSON
  ) where

import Data.Aeson (Value, Result(..), fromJSON, toJSON)
import STUNIR.Types (IRModule)

-- | Parse IR from JSON
parseIR :: Value -> Either String IRModule
parseIR v = case fromJSON v of
  Success m -> Right m
  Error err -> Left err

-- | Convert IR to JSON
irToJSON :: IRModule -> Value
irToJSON = toJSON

File: tools/rust/src/ir.rs
//! IR processing utilities

use crate::types::*;
use anyhow::{Context, Result};
use serde_json::Value;
use std::collections::BTreeMap;

/// Parse IR from JSON
pub fn parse_ir(json: &Value) -> Result<IRModule> {
    serde_json::from_value(json.clone())
        .context("Failed to parse IR module")
}

/// Convert IR module to JSON
pub fn ir_to_json(module: &IRModule) -> Result<Value> {
    serde_json::to_value(module)
        .context("Failed to convert IR to JSON")
}
